### PR TITLE
PHPCS-34 Method docblock summary sniff

### PR DESCRIPTION
**Pull request by nhardeweg**

### CODE REVIEW

File: src/Standards/BestIt/Sniffs/Commenting/MethodDocSniff.php [hunk 1]
```diff
@@ -0,0 +1,234 @@
...
+    /**
+     * A list of tokenizers this sniff supports.
+     *
+     * @var array
+     */
+    public $supportedTokenizers = [
+        'PHP',
+        'JS',
+    ];
+
```
Comment (bestit-tkellner) [inline src/Standards/BestIt/Sniffs/Commenting/MethodDocSniff.php]: Format.

File: src/Standards/BestIt/Sniffs/Commenting/MethodDocSniff.php [hunk 2]
```diff
@@ -0,0 +1,234 @@
...
+
+    /**
+     * Returns an array of tokens this test wants to listen for.
+     *
+     * @return array
+     */
+    public function register()
+    {
+        return array(T_DOC_COMMENT_OPEN_TAG);
```
Comment (bestit-tkellner) [inline src/Standards/BestIt/Sniffs/Commenting/MethodDocSniff.php]: Short array Syntax.

File: src/Standards/BestIt/Sniffs/Commenting/MethodDocSniff.php [hunk 3]
```diff
@@ -0,0 +1,234 @@
...
+     *
+     * @return void
+     */
+    public function process(PHP_CodeSniffer_File $phpcsFile, $stackPtr)
+    {
+        $tokens       = $phpcsFile->getTokens();
+        $commentStart = $stackPtr;
+        $commentEnd   = $tokens[$stackPtr]['comment_closer'];
+
+        $empty = array(
```
Comment (bestit-tkellner) [inline src/Standards/BestIt/Sniffs/Commenting/MethodDocSniff.php]: Short array Syntax.

File: src/Standards/BestIt/Sniffs/Commenting/MethodDocSniff.php [hunk 4]
```diff
@@ -0,0 +1,234 @@
...
+            T_DOC_COMMENT_WHITESPACE,
+            T_DOC_COMMENT_STAR,
+        );
+
+        $short = $phpcsFile->findNext($empty, $stackPtr + 1, $commentEnd, true);
+        if ($short === false) {
+            // No content at all.
+            $error = 'Doc comment is empty';
```
Comment (bestit-tkellner) [inline src/Standards/BestIt/Sniffs/Commenting/MethodDocSniff.php]: Konstante?

File: src/Standards/BestIt/Sniffs/Commenting/MethodDocSniff.php [hunk 5]
```diff
@@ -0,0 +1,234 @@
...
+
+                $phpcsFile->fixer->endChangeset();
+            }
+        }
+
+        // Account for the fact that a short description might cover
+        // multiple lines.
+        $shortContent = $tokens[$short]['content'];
+        $shortEnd     = $short;
+
```
Comment (bestit-tkellner) [inline src/Standards/BestIt/Sniffs/Commenting/MethodDocSniff.php]: Format

**Comment by bestit-tkellner** [comment]
CodeSniffer return an error.

File: README.md [hunk 6]
```diff
@@ -52,6 +52,7 @@ Execute the PHP_CodeSniffer (path can vary on your composer configuration):
 | Generic.Formatting.SpaceAfterCast | BestIt | There MUST be a space after cast. |
 | Generic.Files.LineLength | PSR-2 | Every line MUST not be longer than 120 chars. |
 | Generic.Arrays.DisallowLongArraySyntax | BestIt | Every array syntax MUST be in short array syntax. |
+| MethodDocSniff | BestIt | Every method docblock must have a short description that starts capitalized and is followed by an empty line |
```
Comment (bestit-tkellner) [inline README.md]: You add two rules in xml File but only declare one in Readme

File: tests/Standards/BestIt/Sniffs/Commenting/MethodDocSniffTest.php [hunk 7]
```diff
@@ -0,0 +1,148 @@
...
+use BestIt\Sniffs\Commenting\MethodDocSniff;
+use PHP_CodeSniffer_File;
+use Tests\BestIt\Sniffs\TestCase;
+
+/**
+ * Class FluentSetterSniffTest
+ *
+ * @package Tests\BestIt\Sniffs\Functions
+ *
+ * @author Nick Lubisch <nick.lubisch@bestit-online.de>
```
Comment (bestit-lubisch) [inline tests/Standards/BestIt/Sniffs/Commenting/MethodDocSniffTest.php]: you are the author of this - right? :tongue:

File: tests/Standards/BestIt/Sniffs/Commenting/MethodDocSniffTest.php [hunk 8]
```diff
@@ -0,0 +1,148 @@
...
+declare(strict_types=1);
+
+namespace Tests\BestIt\Sniffs\Commenting;
+
+use BestIt\Sniffs\Commenting\MethodDocSniff;
+use PHP_CodeSniffer_File;
+use Tests\BestIt\Sniffs\TestCase;
+
+/**
+ * Class FluentSetterSniffTest
```
Comment (bestit-lubisch) [inline tests/Standards/BestIt/Sniffs/Commenting/MethodDocSniffTest.php]: class name not correct

File: tests/Standards/BestIt/Sniffs/Commenting/MethodDocSniffTest.php [hunk 9]
```diff
@@ -0,0 +1,148 @@
...
+     */
+    public function testCorrectShortSummary()
+    {
+        $this->assertNoSniffErrorInFile(
+            $this->checkMethodDocSniff(__DIR__ . '/Fixtures/MethodDocSniff.Correct.php')
+        );
+    }
+
+    /**
+     * Test empty method doc block
```
Comment (bestit-lubisch) [inline tests/Standards/BestIt/Sniffs/Commenting/MethodDocSniffTest.php]: return void

File: tests/Standards/BestIt/Sniffs/Commenting/MethodDocSniffTest.php [hunk 10]
```diff
@@ -0,0 +1,148 @@
...
+
+        $this->assertSniffError(
+            $report,
+            5,
+            MethodDocSniff::CODE_EMPTY
+        );
+    }
+
+    /**
+     * Test missing empty line after summary
```
Comment (bestit-lubisch) [inline tests/Standards/BestIt/Sniffs/Commenting/MethodDocSniffTest.php]: return void

File: tests/Standards/BestIt/Sniffs/Commenting/MethodDocSniffTest.php [hunk 11]
```diff
@@ -0,0 +1,148 @@
...
+
+        $this->assertSniffError(
+            $report,
+            27,
+            MethodDocSniff::CODE_SPACING_BETWEEN
+        );
+    }
+
+    /**
+     * Test if a doc block is missing a summary
```
Comment (bestit-lubisch) [inline tests/Standards/BestIt/Sniffs/Commenting/MethodDocSniffTest.php]: return void

File: tests/Standards/BestIt/Sniffs/Commenting/MethodDocSniffTest.php [hunk 12]
```diff
@@ -0,0 +1,148 @@
...
+
+        $this->assertSniffError(
+            $report,
+            5,
+            MethodDocSniff::CODE_MISSING_SHORT
+        );
+    }
+
+    /**
+     * Test if a doc block is missing a summary
```
Comment (bestit-lubisch) [inline tests/Standards/BestIt/Sniffs/Commenting/MethodDocSniffTest.php]: return void

File: tests/Standards/BestIt/Sniffs/Commenting/MethodDocSniffTest.php [hunk 13]
```diff
@@ -0,0 +1,148 @@
...
+
+        $this->assertSniffError(
+            $report,
+            7,
+            MethodDocSniff::CODE_SPACING_BEFORE_SHORT
+        );
+    }
+
+    /**
+     * Tests if the doc block short description (summary) and long description is starting with a capital letter
```
Comment (bestit-lubisch) [inline tests/Standards/BestIt/Sniffs/Commenting/MethodDocSniffTest.php]: return void

File: src/Standards/BestIt/Sniffs/Commenting/MethodDocSniff.php [hunk 14]
```diff
@@ -0,0 +1,234 @@
...
+                $error = 'Doc comment long description must start with a capital letter';
+                $phpcsFile->addError($error, $long, self::CODE_LONG_NOT_CAPITAL);
+            }
+        }
+
+        $long = $phpcsFile->findNext($empty, $shortEnd + 1, $commentEnd - 1, true);
+        if ($long !== false && $tokens[$long]['code'] === T_DOC_COMMENT_STRING) {
+            if ($tokens[$long]['line'] !== ($tokens[$shortEnd]['line'] + 2)) {
+                $error = 'There must be exactly one blank line between descriptions in a doc comment';
+                $fix   = $phpcsFile->addFixableError($error, $long, self::CODE_SPACING_BETWEEN);
```
Comment (bestit-lubisch) [inline src/Standards/BestIt/Sniffs/Commenting/MethodDocSniff.php]: remove identation

File: src/Standards/BestIt/Sniffs/Commenting/MethodDocSniff.php [hunk 15]
```diff
@@ -0,0 +1,234 @@
...
+                }
+
+                $phpcsFile->fixer->endChangeset();
+            }
+        }
+
+        // Account for the fact that a short description might cover
+        // multiple lines.
+        $shortContent = $tokens[$short]['content'];
+        $shortEnd     = $short;
```
Comment (bestit-lubisch) [inline src/Standards/BestIt/Sniffs/Commenting/MethodDocSniff.php]: remove identation

File: src/Standards/BestIt/Sniffs/Commenting/MethodDocSniff.php [hunk 16]
```diff
@@ -0,0 +1,234 @@
...
+        if ($tokens[$short]['code'] !== T_DOC_COMMENT_STRING) {
+            $error = 'Missing short description in doc comment';
+            $phpcsFile->addError($error, $stackPtr, self::CODE_MISSING_SHORT);
+            return;
+        }
+
+        // No extra newline before short description.
+        if ($tokens[$short]['line'] !== ($tokens[$stackPtr]['line'] + 1)) {
+            $error = 'Doc comment short description must be on the first line';
+            $fix   = $phpcsFile->addFixableError($error, $short, self::CODE_SPACING_BEFORE_SHORT);
```
Comment (bestit-lubisch) [inline src/Standards/BestIt/Sniffs/Commenting/MethodDocSniff.php]: remove identation

File: src/Standards/BestIt/Sniffs/Commenting/MethodDocSniff.php [hunk 17]
```diff
@@ -0,0 +1,234 @@
...
+            return;
+        }
+
+        // The last line of the comment should just be the */ code.
+        $prev = $phpcsFile->findPrevious($empty, $commentEnd - 1, $stackPtr, true);
+
+        // Check for additional blank lines at the end of the comment.
+        if ($tokens[$prev]['line'] < ($tokens[$commentEnd]['line'] - 1)) {
+            $error = 'Additional blank lines found at end of doc comment';
+            $fix   = $phpcsFile->addFixableError($error, $commentEnd, 'SpacingAfter');
```
Comment (bestit-lubisch) [inline src/Standards/BestIt/Sniffs/Commenting/MethodDocSniff.php]: remove identation

File: src/Standards/BestIt/Sniffs/Commenting/MethodDocSniff.php [hunk 18]
```diff
@@ -0,0 +1,234 @@
...
+     * Processes this test, when one of its tokens is encountered.
+     *
+     * @param PHP_CodeSniffer_File $phpcsFile The file being scanned.
+     * @param int $stackPtr  The position of the current token in the stack passed in $tokens.
+     *
+     * @return void
+     */
+    public function process(PHP_CodeSniffer_File $phpcsFile, $stackPtr)
+    {
+        $tokens       = $phpcsFile->getTokens();
```
Comment (bestit-lubisch) [inline src/Standards/BestIt/Sniffs/Commenting/MethodDocSniff.php]: remove identation

File: src/Standards/BestIt/Sniffs/Commenting/MethodDocSniff.php [hunk 19]
```diff
@@ -0,0 +1,234 @@
...
+    public function register()
+    {
+        return array(T_DOC_COMMENT_OPEN_TAG);
+    }
+
+    /**
+     * Processes this test, when one of its tokens is encountered.
+     *
+     * @param PHP_CodeSniffer_File $phpcsFile The file being scanned.
+     * @param int $stackPtr  The position of the current token in the stack passed in $tokens.
```
Comment (bestit-lubisch) [inline src/Standards/BestIt/Sniffs/Commenting/MethodDocSniff.php]: unnesseccary double space after $stackPtr

File: README.md (hunk 6 above)
Comment (nhardeweg) [inline README.md]: Removed the second Rule

File: src/Standards/BestIt/Sniffs/Commenting/MethodDocSniff.php [hunk 20]
```diff
@@ -0,0 +1,234 @@
...
+
+use PHP_CodeSniffer_File;
+use PHP_CodeSniffer_Sniff;
+
+/**
+ * This Sniff checks the method phpdoc
+ *
+ * It is a modified version of the Generic DocCommentSniff adjusted for the bestit codestyle
+ *
+ * @package BestIt\Sniffs\Commenting
```
Comment (nhardeweg) [inline src/Standards/BestIt/Sniffs/Commenting/MethodDocSniff.php]: Done

File: tests/Standards/BestIt/Sniffs/Commenting/MethodDocSniffTest.php (hunk 9 above)
Comment (nhardeweg) [inline tests/Standards/BestIt/Sniffs/Commenting/MethodDocSniffTest.php]: Done

File: tests/Standards/BestIt/Sniffs/Commenting/MethodDocSniffTest.php (hunk 10 above)
Comment (nhardeweg) [inline tests/Standards/BestIt/Sniffs/Commenting/MethodDocSniffTest.php]: Done

File: tests/Standards/BestIt/Sniffs/Commenting/MethodDocSniffTest.php (hunk 11 above)
Comment (nhardeweg) [inline tests/Standards/BestIt/Sniffs/Commenting/MethodDocSniffTest.php]: Done

File: tests/Standards/BestIt/Sniffs/Commenting/MethodDocSniffTest.php (hunk 12 above)
Comment (nhardeweg) [inline tests/Standards/BestIt/Sniffs/Commenting/MethodDocSniffTest.php]: Done

File: tests/Standards/BestIt/Sniffs/Commenting/MethodDocSniffTest.php (hunk 13 above)
Comment (nhardeweg) [inline tests/Standards/BestIt/Sniffs/Commenting/MethodDocSniffTest.php]: Done

File: src/Standards/BestIt/Sniffs/Commenting/MethodDocSniff.php [hunk 21]
```diff
@@ -0,0 +1,293 @@
...
+
+    /**
+     * Processes this test, when one of its tokens is encountered.
+     *
+     * @param PHP_CodeSniffer_File $phpcsFile The file being scanned.
+     * @param int $stackPtr The position of the current token in the stack passed in $tokens.
+     *
+     * @return void
+     */
+    public function process(PHP_CodeSniffer_File $phpcsFile, $stackPtr)
```
Comment (Haehnchen) [inline src/Standards/BestIt/Sniffs/Commenting/MethodDocSniff.php]: how do you handle inline decls?
```
/** @var $foo Foobar **/`
/** @var Foobar $foo **/`
```

whats about add a _nextSibling_ check for a _method token_ in doc comment scope?

File: src/Standards/BestIt/Sniffs/Commenting/MethodDocSniff.php [hunk 22]
```diff
@@ -0,0 +1,293 @@
...
+                        break;
+                    }
+                    $phpcsFile->fixer->replaceToken($i, '');
+                }
+                $phpcsFile->fixer->endChangeset();
+            }
+        }
+
+        // Check for a comment description.
+        if ($tokens[$short]['code'] !== T_DOC_COMMENT_STRING && $tokens[$short]['content'] !== '@inheritdoc') {
```
Comment (Haehnchen) [inline src/Standards/BestIt/Sniffs/Commenting/MethodDocSniff.php]: `{@inheritdoc}` is also valid to use

File: tests/Standards/BestIt/Sniffs/Commenting/MethodDocSniffTest.php [hunk 23]
```diff
@@ -0,0 +1,160 @@
...
+use Tests\BestIt\Sniffs\TestCase;
+
+/**
+ * Class MethodDocSniffTest
+ *
+ * @package Tests\BestIt\Sniffs\Functions
+ *
+ * @author Nils Hardeweg <nils.hardeweg@bestit-online.de>
+ */
+class MethodDocSniffTest extends TestCase
```
Comment (Haehnchen) [inline tests/Standards/BestIt/Sniffs/Commenting/MethodDocSniffTest.php]: is there no way for negative test like `@inheritdoc`?

Comment (nhardeweg) [inline tests/Standards/BestIt/Sniffs/Commenting/MethodDocSniffTest.php]: Test for added to cover @inheritdoc

File: src/Standards/BestIt/Sniffs/Commenting/MethodDocSniff.php (hunk 21 above)
Comment (nhardeweg) [inline src/Standards/BestIt/Sniffs/Commenting/MethodDocSniff.php]: Inline declarations are accounted for. If there is @var in the first line the sniff will not throw an error.
This is checked via unittest

File: src/Standards/BestIt/Sniffs/Commenting/MethodDocSniff.php [hunk 24]
```diff
@@ -0,0 +1,367 @@
...
+        $prev = $this->phpcsFile->findPrevious($this->empty, $commentEnd - 1, $stackPtr, true);
+
+        $this->checkBlankLines($commentEnd, $prev);
+
+        // Check for a comment description.
+        if ($this->tokens[$this->short]['code'] !== T_DOC_COMMENT_STRING
+            && $this->tokens[$this->short]['content'] !== '@inheritdoc'
+            && $this->tokens[$this->short]['content'] !== '@var'
+        ) {
+            $error = self::MESSAGE_MISSING_SHORT;
```
Comment (bestit-lubisch) [inline src/Standards/BestIt/Sniffs/Commenting/MethodDocSniff.php]: directly use as first parameter of addError

File: src/Standards/BestIt/Sniffs/Commenting/MethodDocSniff.php [hunk 25]
```diff
@@ -0,0 +1,367 @@
...
+     * @param $stackPtr
+     * @param $commentEnd
+     *
+     * @return void
+     */
+    private function checkSummaryAndLongDescription($stackPtr, $commentEnd)
+    {
+        // No extra newline before short description.
+        if ($this->tokens[$this->short]['content'] !== '@var'
+            && $this->tokens[$this->short]['line'] !== ($this->tokens[$stackPtr]['line'] + 1)) {
```
Comment (bestit-lubisch) [inline src/Standards/BestIt/Sniffs/Commenting/MethodDocSniff.php]: ") {" into new line

File: src/Standards/BestIt/Sniffs/Commenting/MethodDocSniff.php [hunk 26]
```diff
@@ -0,0 +1,367 @@
...
+     *
+     * @param $stackPtr
+     *
+     * @return void
+     */
+    private function checkCommentTags($stackPtr)
+    {
+        if (!empty($this->tokens[$this->commentStart]['comment_tags'])) {
+            $firstTag = $this->tokens[$this->commentStart]['comment_tags'][0];
+            $prev = $this->phpcsFile->findPrevious($this->empty, $firstTag - 1, $stackPtr, true);
```
Comment (bestit-lubisch) [inline src/Standards/BestIt/Sniffs/Commenting/MethodDocSniff.php]: previous what?

File: src/Standards/BestIt/Sniffs/Commenting/MethodDocSniff.php [hunk 27]
```diff
@@ -0,0 +1,367 @@
...
+     *
+     * @param $stackPtr
+     * @param $commentEnd
+     *
+     * @return void
+     */
+    private function checkMethodPhpDoc($stackPtr, $commentEnd)
+    {
+        // The last line of the comment should just be the */ code.
+        $prev = $this->phpcsFile->findPrevious($this->empty, $commentEnd - 1, $stackPtr, true);
```
Comment (bestit-lubisch) [inline src/Standards/BestIt/Sniffs/Commenting/MethodDocSniff.php]: previous what?

File: README.md [hunk 28]
```diff
@@ -54,6 +54,7 @@ Execute the PHP_CodeSniffer (path can vary on your composer configuration):
 | Generic.Files.LineLength | PSR-2 | Every line MUST not be longer than 120 chars. |
 | Generic.Arrays.DisallowLongArraySyntax | BestIt | Every array syntax MUST be in short array syntax. |
 | BestIt.Formatting.OpenTagSniff | BestIt | After the open tag there MUST be an empty line. |
+| MethodDocSniff | BestIt | Every method docblock MUST have a short description that starts capitalized and is followed by an empty line |
```
Comment (bestit-lubisch) [inline README.md]: Full name in first column

**Pull request merged**